sled-agent/src/services.rs

```diff
@@ -2837,6 +2888,17 @@ impl ServiceManager {
                 }
             }
 
+            // For each new zone request, ensure that we've sufficiently
+            // synchronized time.
+            //
+            // NOTE: This imposes a constraint, during initial setup, cold boot,
+            // etc, that NTP and the internal DNS system it depends on MUST be
+            // initialized prior to other zones.
+            if zone_requires_timesync(&zone.zone_type) && !time_is_synchronized
```

Comment:
The PR looks good but I'm not sure about the check here.  (I don't know that it's wrong; I just ran out of time right now to convince myself it's right.)  My two concerns are:

1. This has a lot more callers than `ensure_all_omicron_zones_persistent()`.  But I just looked and the only ones _outside_ of `ensure_all_omicron_zones_persistent()` are all in `load_services()`, where it seems fairly clear that this will be okay because we've already waited for timesync at the right spot.  So I think this is okay.
2. If we do fail at this point, we may have already destroyed zones that aren't supposed to be running in the new generation, but we're also rejecting the new generation.  I know this falls in the category of "side effects" that we talked about earlier and said we might ignore for now, but this seems a lot worse than the side effects we were talking about before, where we just created an extra dataset.  In the other case, the system is still running what it's supposed to be running.  In this case, it wouldn't be.  Right?

Reply:
> 1. This has a lot more callers than `ensure_all_omicron_zones_persistent()`.  But I just looked and the only ones _outside_ of `ensure_all_omicron_zones_persistent()` are all in `load_services()`, where it seems fairly clear that this will be okay because we've already waited for timesync at the right spot.  So I think this is okay.

Yeah, if anything, if we do want this check to prevent zones from truly loading before timesync, putting it on the "common" path seemed better.

> 2. If we do fail at this point, we may have already destroyed zones that aren't supposed to be running in the new generation, but we're also rejecting the new generation.  I know this falls in the category of "side effects" that we talked about earlier and said we might ignore for now, but this seems a lot worse than the side effects we were talking about before, where we just created an extra dataset.  In the other case, the system is still running what it's supposed to be running.  In this case, it wouldn't be.  Right?

I somewhat agree, that's why I initially put the check in a much higher-level location to bail out early. However, I rationalized this to myself with the following: the state of "we destroyed the zones that should not be here, and we are getting ready to make the new ones" is absolutely a valid intermediate state while performing this operation. If we fail, the endpoint can be called again (or `load_services` can be re-invoked) and we'll proceed as we should.

So in that sense: is there really *value* in leaving the old zones up? A new request arrived asking for them to be destroyed, and that would be a valid intermediate state. I'd argue that this situation is actually not that distinct from "You had zones {A, B, C}, you now ask for zones {D, E, F}, and after destroying {A, B, C}, we fail to create {D} for any reason". That's how the code is currently implemented, regardless of this PR, and I think this change is consistent with those guarantees.

Note that if a zone was in both the "old" and "new" config, it'll be left alone.

Reply:
Hmm.  I hear what you're saying.  I agree these are all problematic and we probably want to move toward a model like we do with RPWs where we validate that the request is logically sound (i.e., that we _should_ be able to apply it), accept it, and then retry indefinitely to make it happen.  And that's outside the scope of what we're doing here.

This new case feels a little different to me than the existing ones.  For the existing cases where sled agent could fail to create a zone, they seem unlikely, unknowable ahead of time (so not easily avoidable), and either transient (e.g., out of memory) or unrecoverable (e.g., some step went haywire and needs human attention).  This seems like a problem but not an urgent one.  (Have we seen cases like that outside of development?)  On the other hand, the condition we're adding here is likely (we're intending to use it as part of normal operation) so it seems pragmatically worse -- but it's also knowable up front so easy to avoid.

Oh, how about this alternative: keep the check inside this function, but do it first thing, before having made any changes?  I think that preserves the upside (keeping it in common code) but without the risk?

Reply:
> Hmm. I hear what you're saying. I agree these are all problematic and we probably want to move toward a model like we do with RPWs where we validate that the request is logically sound (i.e., that we _should_ be able to apply it), accept it, and then retry indefinitely to make it happen. And that's outside the scope of what we're doing here.

I agree with this direction -- IMO this is basically the approach we'd take to "async-ify" the endpoint -- but I'm less certain we could ever fully validate that all the zones will successfully launch without actively trying to launch them. We can certainly try to catch obviously broken configs, but I think it's a difficult task (and not a necessary one!) to ensure that "once we accept a request to launch a certain set of services, we **must** guarantee that they'll be launch-able".

For example: Someone could ask for a zone using a dataset on a U.2 disk that gets unplugged! We'll always need to cope with "some set of zones may not be runnable".

> This new case feels a little different to me than the existing ones. For the existing cases where sled agent could fail to create a zone, they seem unlikely, unknowable ahead of time (so not easily avoidable), and either transient (e.g., out of memory) or unrecoverable (e.g., some step went haywire and needs human attention). This seems like a problem but not an urgent one. (Have we seen cases like that outside of development?) On the other hand, the condition we're adding here is likely (we're intending to use it as part of normal operation) so it seems pragmatically worse -- but it's also knowable up front so easy to avoid.

I disagree with the assertion that this is pragmatically worse! I think this is fundamentally a transient condition, like out-of-memory. Yes, it has slightly different behavior, in the sense that "we don't **expect** this transient failure to re-appear after the initial timesync has completed", but it's basically a transient case for someone making this request at an HTTP level. 

Also, I'm not sure it being "knowable ahead-of-time" necessarily forces our hand into making the operation avoid all side-effects. We don't even try to validate most of the zone configuration (in the form of both addresses, free-form strings as "domains", etc), and I think that's okay because whether or not the information can be validated ahead-of-time is only for convenience and error message clarity, and it's not necessary for correctness.

> Oh, how about this alternative: keep the check inside this function, but do it first thing, before having made any changes? I think that preserves the upside (keeping it in common code) but without the risk?

If we did this, we'd still be causing side-effects, due to the dataset generation I described in https://github.com/oxidecomputer/omicron/pull/4778#issuecomment-1881743613 .

I've moved it regardless, so now the logic is basically:

- Find zones to {add, remove}
- Do some checks on zones to be added <-- Timesync check now here
- Remove zones to be removed
- Do more checks on zones to be added
- Add zones to be added